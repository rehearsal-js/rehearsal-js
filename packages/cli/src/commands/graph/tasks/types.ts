export interface GraphWorkerOptions {
  rootPath: string;
  srcPath: string;
  ignore: string[];
  serviceMap: Record<string, string>;
  output?: string;
}

type MessageResponse = { type: 'message'; content: string };
type FilesResponse = { type: 'files'; content: string[] };
export type Response = MessageResponse | FilesResponse;
