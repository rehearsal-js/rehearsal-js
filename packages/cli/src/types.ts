import { MigrationStrategy } from '@rehearsal/migration-graph';

import { UserConfig } from './userConfig';
import { State } from './helpers/state';

export type CliCommand = 'upgrade' | 'migrate';
export type Formats = 'sarif' | 'json' | 'sonarqube' | 'md';

export type MigrateCommandOptions = {
  basePath: string;
  entrypoint: string;
  files: string;
  format: Formats[];
  outputPath: string;
  verbose: boolean | undefined;
  userConfig: string | undefined;
  interactive: boolean | undefined;
};

export type MigrateCommandContext = {
  skip: boolean;
  userConfig: UserConfig | undefined;
  strategy: MigrationStrategy;
  sourceFilesWithAbsolutePath: string[];
  sourceFilesWithRelativePath: string[];
  input: unknown;
  targetPackagePath: string;
  state: State;
};

export type UpgradeCommandContext = {
  tsVersion: string;
  latestAvailableBuild: string;
  currentTSVersion: string;
  skip: boolean;
};

export type UpgradeCommandOptions = {
  build: string;
  basePath: string;
  format: Formats[];
  outputPath: string;
  tsVersion: string;
  dryRun: boolean;
};

export type DependencyConfig = {
  dependencies?: string[];
  devDependencies?: string[];
};

export type SetupConfigCommand = {
  command: string;
  args: string[];
};

export type SetupConfig = {
  ts?: SetupConfigCommand;
  lint?: SetupConfigCommand;
};

export type CustomCommandConfig = {
  install?: DependencyConfig;
  setup?: SetupConfig;
};

export type CustomConfig = {
  upgrade?: CustomCommandConfig;
  migrate?: CustomCommandConfig;
};

export type PackageSelection = {
  name: string;
  path: string;
};

export type MenuMap = {
  [key: string]: string;
};

export type MigrationSummary = {
  totalErrorCount: number;
  hintAddedCount: number;
};
export type TSConfig = {
  compilerOptions: {
    strict: boolean;
  };
};

export type ScriptMap = {
  [key: string]: string;
};
