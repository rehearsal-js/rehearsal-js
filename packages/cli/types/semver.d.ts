declare module 'semver' {
  const semver: {
    valid: (input: string) => boolean;
  };
  export = semver;
}
