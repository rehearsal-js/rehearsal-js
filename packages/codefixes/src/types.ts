import {
  CodeFixAction,
  DiagnosticWithLocation,
  FileTextChanges,
  LanguageService,
  Node,
  Program,
  TypeChecker,
} from 'typescript';

export interface DiagnosticWithContext extends DiagnosticWithLocation {
  service: LanguageService;
  program: Program;
  checker: TypeChecker;
  node?: Node;
}

export interface CodeHintList {
  [key: number]: CodeHintListElement;
}

export interface CodeHintListElement {
  hint: string;
  helpUrl?: string;
  hints?: CodeHint[];
}

export interface CodeHint {
  when: (n: Node, p: Program, c: TypeChecker) => boolean;
  hint: string;
}

export interface CodeFixCollection {
  getFixForDiagnostic(diagnostic: DiagnosticWithContext): CodeFixAction | undefined;
}

export type CodeFixKind = 'insert' | 'delete' | 'replace';

export interface FixedFile {
  fileName: string;
  newCode?: string;
  oldCode?: string;
  codeFixAction?: CodeFixKind;
  location: {
    startLine: number;
    startColumn: number;
    endLine: number;
    endColumn: number;
  };
}

export interface CodeFix {
  getCodeAction: (diagnostic: DiagnosticWithContext) => CodeFixAction | undefined;
}

export function createCodeFixAction(
  fixName: string,
  changes: FileTextChanges[],
  description: string
): CodeFixAction {
  return {
    fixName,
    description,
    changes,
  };
}
