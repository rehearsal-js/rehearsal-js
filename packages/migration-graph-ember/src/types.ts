/* eslint-disable @typescript-eslint/no-explicit-any */
import type { Package } from '@rehearsal/migration-graph-shared';
import type { EmberAddonPackage } from './entities/ember-addon-package.js';
import type { EmberAppPackage } from './entities/ember-app-package.js';

export type EmberProjectPackage = Package | EmberAddonPackage | EmberAppPackage;
