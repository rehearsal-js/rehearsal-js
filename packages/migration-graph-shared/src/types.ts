import type { Package } from './entities/package.js';

export type UniqueNode = {
  key: string;
  synthetic?: boolean;
};

export type PackageNode = UniqueNode & {
  pkg: Package;
  converted?: boolean;
};

export type ModuleNode = UniqueNode & {
  path: string;
  parsed?: boolean;
};
