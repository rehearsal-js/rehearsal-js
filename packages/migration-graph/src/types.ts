import { Package } from '@rehearsal/migration-graph-shared';

import { Graph } from './utils/graph';

export type UniqueGraphNode = {
  key: string;
};

export type PackageNode = UniqueGraphNode & {
  pkg: Package;
  converted?: boolean;
  modules: Graph<ModuleNode>;
};

export type ModuleNode = UniqueGraphNode & {
  path: string;
  parsed?: boolean;
  meta?: unknown;
};
