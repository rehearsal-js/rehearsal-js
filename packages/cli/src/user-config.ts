import { resolve } from 'node:path';
import { addDep } from '@rehearsal/utils';
import { execa } from 'execa';
import { readJSONSync } from 'fs-extra/esm';
import type {
  MigrateCommandConfig,
  UpgradeCommandConfig,
  CliCommand,
  CustomConfig,
} from './types.js';

// Storage and runner for user custom cli config
export class UserConfig {
  public basePath: string;
  public config?: MigrateCommandConfig | UpgradeCommandConfig;

  constructor(basePath: string, command: CliCommand) {
    const config: CustomConfig = readJSONSync(resolve(basePath, 'rehearsal-config.json'));

    this.config = config[command];

    this.basePath = basePath;
  }

  public get hasDependencies(): boolean {
    return (
      !!this.config?.install &&
      (!!this.config?.install?.dependencies || !!this.config?.install?.devDependencies)
    );
  }

  public get dependencies(): string[] {
    if (this.config && this.config.install) {
      const { dependencies } = this.config.install;
      if (dependencies && dependencies.length) {
        return dependencies;
      }
    }
    return [];
  }

  public get devDependencies(): string[] {
    if (this.config && this.config.install) {
      const { devDependencies } = this.config.install;
      if (devDependencies && devDependencies.length) {
        return devDependencies;
      }
    }
    return [];
  }

  public get hasPostInstallHook(): boolean {
    return !!this.config?.postInstall;
  }

  public get hasTsSetup(): boolean {
    return !!this.config?.setup?.ts;
  }

  public get hasPostTsSetupHook(): boolean {
    return !!this.config?.setup?.postTsSetup;
  }

  public get hasLintSetup(): boolean {
    return !!this.config?.setup?.lint;
  }

  public get hasPostLintSetup(): boolean {
    return !!this.config?.setup?.postLintSetup;
  }

  async install(): Promise<void> {
    if (this.config && this.config.install) {
      const { dependencies, devDependencies } = this.config.install;
      if (dependencies && dependencies.length) {
        await addDep(dependencies, false, { cwd: this.basePath });
      }

      if (devDependencies && devDependencies.length) {
        await addDep(devDependencies, true, { cwd: this.basePath });
      }
    }
  }

  async postInstall(): Promise<void> {
    if (this.config && this.config.postInstall) {
      if (Array.isArray(this.config.postInstall)) {
        for (const { command, args } of this.config.postInstall) {
          await execa(command, args, { cwd: this.basePath });
        }
      } else {
        const { command, args } = this.config.postInstall;
        await execa(command, args, { cwd: this.basePath });
      }
    }
  }

  // TODO: how to make setup tasks more generic instead of separated functions?

  // TODO: it doesn't make sense for migrate and upgrade sharing the same setup tasks,
  // What does upgrade command need for custom config?
  async tsSetup(): Promise<void> {
    if (this.config?.setup?.ts) {
      const { ts: tsSetup } = this.config.setup;
      if (Array.isArray(tsSetup)) {
        for (const { command, args } of tsSetup) {
          await execa(command, args, { cwd: this.basePath });
        }
      } else {
        const { command, args } = tsSetup;
        await execa(command, args, { cwd: this.basePath });
      }
    }
  }

  async postTsSetup(): Promise<void> {
    if (this.config?.setup?.postTsSetup) {
      const { postTsSetup } = this.config.setup;
      if (Array.isArray(postTsSetup)) {
        for (const { command, args } of postTsSetup) {
          await execa(command, args, { cwd: this.basePath });
        }
      } else {
        const { command, args } = postTsSetup;
        await execa(command, args, { cwd: this.basePath });
      }
    }
  }

  async lintSetup(): Promise<void> {
    if (this.config?.setup?.lint) {
      const { lint: lintSetup } = this.config.setup;
      if (Array.isArray(lintSetup)) {
        for (const { command, args } of lintSetup) {
          await execa(command, args, { cwd: this.basePath });
        }
      } else {
        const { command, args } = lintSetup;
        await execa(command, args, { cwd: this.basePath });
      }
    }
  }

  async postLintSetup(): Promise<void> {
    if (this.config?.setup?.postLintSetup) {
      const { postLintSetup } = this.config.setup;
      if (Array.isArray(postLintSetup)) {
        for (const { command, args } of postLintSetup) {
          await execa(command, args, { cwd: this.basePath });
        }
      } else {
        const { command, args } = postLintSetup;
        await execa(command, args, { cwd: this.basePath });
      }
    }
  }

  get exclude(): string[] | undefined {
    return (this.config as MigrateCommandConfig)?.exclude;
  }

  get include(): string[] | undefined {
    return (this.config as MigrateCommandConfig)?.include;
  }
}
