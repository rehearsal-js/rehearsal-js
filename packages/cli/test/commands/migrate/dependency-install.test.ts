import { resolve } from 'node:path';
import { existsSync, promises as fs } from 'node:fs';
import { afterEach, beforeEach, describe, expect, test, vi } from 'vitest';
import { writeJSONSync } from 'fs-extra/esm';

import {
  REQUIRED_DEPENDENCIES,
  depInstallTask,
  shouldRunDepInstallTask,
} from '../../../src/commands/migrate/tasks/index.js';
import { prepareTmpDir, listrTaskRunner, createMigrateOptions } from '../../test-helpers/index.js';
import { CustomConfig } from '../../../src/types.js';
import { UserConfig } from '../../../src/user-config.js';
import type { PackageJson } from 'type-fest';

function createUserConfig(basePath: string, config: CustomConfig): void {
  const configPath = resolve(basePath, 'rehearsal-config.json');
  writeJSONSync(configPath, config);
}

describe('Task: dependency-install', () => {
  let basePath = '';
  let output = '';
  vi.spyOn(console, 'info').mockImplementation((chunk) => {
    output += `${chunk}\n`;
  });
  vi.spyOn(console, 'log').mockImplementation((chunk) => {
    output += `${chunk}\n`;
  });

  vi.spyOn(console, 'error').mockImplementation((chunk) => {
    output += `${chunk}\n`;
  });

  beforeEach(() => {
    output = '';
    basePath = prepareTmpDir('initialization');
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  test('install required dependencies', async () => {
    const options = createMigrateOptions(basePath);
    const tasks = [depInstallTask(options)];
    await listrTaskRunner(tasks);

    const packageJson = JSON.parse(
      await fs.readFile(resolve(basePath, 'package.json'), 'utf-8')
    ) as PackageJson;

    const devDeps = packageJson.devDependencies;

    expect(Object.keys(devDeps || {}).sort()).toEqual(REQUIRED_DEPENDENCIES.sort());
    expect(output).matchSnapshot();
  });

  test('shouldRunDepInstallTask should skip', async () => {
    const options = createMigrateOptions(basePath);

    // convert array of deps to object-ish
    const requiredDevDepsMap = REQUIRED_DEPENDENCIES.reduce(
      (map, d) => ({ ...map, [d]: '1.0.0' }),
      {}
    );
    // update package.json with required deps
    const packageJsonPath = resolve(basePath, 'package.json');
    const packageJson = JSON.parse(await fs.readFile(packageJsonPath, 'utf-8')) as PackageJson;
    writeJSONSync(packageJsonPath, {
      ...packageJson,
      devDependencies: requiredDevDepsMap,
    });

    expect(await shouldRunDepInstallTask(options)).toBeFalsy();
  });

  test('skip install required dependencies', async () => {
    const options = createMigrateOptions(basePath);
    const tasks = [depInstallTask(options)];

    // convert array of deps to object-ish
    const requiredDevDepsMap = REQUIRED_DEPENDENCIES.reduce(
      (map, d) => ({ ...map, [d]: '1.0.0' }),
      {}
    );
    // update package.json with required deps
    const packageJsonPath = resolve(basePath, 'package.json');
    const packageJson = JSON.parse(await fs.readFile(packageJsonPath, 'utf-8')) as PackageJson;
    writeJSONSync(packageJsonPath, {
      ...packageJson,
      devDependencies: requiredDevDepsMap,
    });

    await listrTaskRunner(tasks); // should be skipped
    expect(output).toMatchSnapshot();
  });

  test('install custom dependencies', async () => {
    createUserConfig(basePath, {
      migrate: {
        install: {
          dependencies: ['fs-extra'],
          devDependencies: ['tmp'],
        },
      },
    });
    const userConfig = new UserConfig(basePath, 'rehearsal-config.json', 'migrate');
    const options = createMigrateOptions(basePath, { userConfig: 'rehearsal-config.json' });
    const tasks = [depInstallTask(options, { userConfig })];
    await listrTaskRunner(tasks);

    const packageJson = JSON.parse(
      await fs.readFile(resolve(basePath, 'package.json'), 'utf-8')
    ) as PackageJson;
    const devDeps = packageJson.devDependencies;
    const deps = packageJson.dependencies;

    expect(Object.keys(devDeps || {}).sort()).toEqual(['tmp', ...REQUIRED_DEPENDENCIES].sort());

    expect(devDeps).toHaveProperty('tmp');
    expect(deps).toHaveProperty('fs-extra');
    expect(output).matchSnapshot();
  });

  test('shouldRunDepInstallTask should skip with custom deps', async () => {
    createUserConfig(basePath, {
      migrate: {
        install: {
          dependencies: ['fs-extra@2.0.0'], // also handle this format: {name}@{version}
          devDependencies: ['tmp'],
        },
      },
    });
    const userConfig = new UserConfig(basePath, 'rehearsal-config.json', 'migrate');
    const options = createMigrateOptions(basePath, { userConfig: 'rehearsal-config.json' });
    // convert array of deps to object-ish
    const requiredDevDepsMap = {
      ...REQUIRED_DEPENDENCIES.reduce((map, d) => ({ ...map, [d]: '1.0.0' }), {}),
      tmp: '1.0.0',
    };
    const requiredDepsMap = { 'fs-extra': '2.0.0' };
    // update package.json with required deps
    const packageJsonPath = resolve(basePath, 'package.json');
    const packageJson = JSON.parse(await fs.readFile(packageJsonPath, 'utf-8')) as PackageJson;
    writeJSONSync(packageJsonPath, {
      ...packageJson,
      dependencies: requiredDepsMap,
      devDependencies: requiredDevDepsMap,
    });
    expect(await shouldRunDepInstallTask(options, { userConfig })).toBeFalsy();
  });

  test('skip install custom deps', async () => {
    createUserConfig(basePath, {
      migrate: {
        install: {
          dependencies: ['fs-extra@2.0.0'], // also handle this format: {name}@{version}
          devDependencies: ['tmp'],
        },
      },
    });
    const userConfig = new UserConfig(basePath, 'rehearsal-config.json', 'migrate');
    const options = createMigrateOptions(basePath, { userConfig: 'rehearsal-config.json' });
    const tasks = [depInstallTask(options, { userConfig })];
    // convert array of deps to object-ish
    const requiredDevDepsMap = {
      ...REQUIRED_DEPENDENCIES.reduce((map, d) => ({ ...map, [d]: '1.0.0' }), {}),
      tmp: '1.0.0',
    };
    const requiredDepsMap = { 'fs-extra': '2.0.0' };
    // update package.json with required deps
    const packageJsonPath = resolve(basePath, 'package.json');
    const packageJson = JSON.parse(await fs.readFile(packageJsonPath, 'utf-8')) as PackageJson;
    writeJSONSync(packageJsonPath, {
      ...packageJson,
      dependencies: requiredDepsMap,
      devDependencies: requiredDevDepsMap,
    });

    await listrTaskRunner(tasks); // should be skipped
    expect(output).toMatchSnapshot();
  });

  test('show failed deps', async () => {
    createUserConfig(basePath, {
      migrate: {
        install: {
          dependencies: ['this-is-not-existed'],
          devDependencies: ['this-would-fail'],
        },
      },
    });
    const userConfig = new UserConfig(basePath, 'rehearsal-config.json', 'migrate');
    const options = createMigrateOptions(basePath, { userConfig: 'rehearsal-config.json' });
    const tasks = [depInstallTask(options, { userConfig })];
    // await listrTaskRunner(tasks);

    await expect(() => listrTaskRunner(tasks)).rejects.toThrowErrorMatchingSnapshot();
  });

  test('single postInstall command from user config', async () => {
    createUserConfig(basePath, {
      migrate: {
        install: {
          dependencies: [],
          devDependencies: ['fs-extra'],
        },
        postInstall: {
          command: 'touch',
          args: ['foo'],
        },
      },
    });
    const userConfig = new UserConfig(basePath, 'rehearsal-config.json', 'migrate');
    const options = createMigrateOptions(basePath, { userConfig: 'rehearsal-config.json' });
    const tasks = [depInstallTask(options, { userConfig })];
    await listrTaskRunner(tasks);

    const packageJson = JSON.parse(
      await fs.readFile(resolve(basePath, 'package.json'), 'utf-8')
    ) as PackageJson;
    const devDeps = packageJson.devDependencies;
    expect(devDeps).toHaveProperty('fs-extra');

    expect(existsSync(resolve(basePath, 'foo')));
    expect(output).matchSnapshot();
  });

  test('multiple postInstall commands from user config', async () => {
    createUserConfig(basePath, {
      migrate: {
        install: {
          dependencies: [],
          devDependencies: [],
        },
        postInstall: [
          {
            command: 'touch',
            args: ['foo'],
          },
          {
            command: 'touch',
            args: ['bar'],
          },
        ],
      },
    });
    const userConfig = new UserConfig(basePath, 'rehearsal-config.json', 'migrate');
    const options = createMigrateOptions(basePath, { userConfig: 'rehearsal-config.json' });
    const tasks = [depInstallTask(options, { userConfig })];
    await listrTaskRunner(tasks);

    expect(existsSync(resolve(basePath, 'foo')));
    expect(existsSync(resolve(basePath, 'bar')));
    expect(output).matchSnapshot();
  });

  test('do not run postInstall command if empty', async () => {
    createUserConfig(basePath, {
      migrate: {
        install: {
          dependencies: [],
          devDependencies: [],
        },
        postInstall: [],
      },
    });
    const userConfig = new UserConfig(basePath, 'rehearsal-config.json', 'migrate');
    const options = createMigrateOptions(basePath, { userConfig: 'rehearsal-config.json' });
    const tasks = [depInstallTask(options, { userConfig })];
    await listrTaskRunner(tasks);

    expect(output).matchSnapshot();
  });
});
