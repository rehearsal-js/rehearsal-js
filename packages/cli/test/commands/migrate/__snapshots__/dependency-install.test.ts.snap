// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Task: dependency-install > do not run postInstall command if empty 1`] = `
"[STARTED] Install dependencies
[DATA] Installing devDependencies: @types/node, @typescript-eslint/eslint-plugin, @typescript-eslint/parser, eslint, eslint-config-prettier, eslint-plugin-prettier, prettier, typescript
[DATA] Setting resolutions: typescript@^5.0.4
[SUCCESS] Install dependencies
"
`;

exports[`Task: dependency-install > install custom dependencies 1`] = `
"[STARTED] Install dependencies
[DATA] Installing dependencies: fs-extra
[DATA] Installing devDependencies: @types/node, @typescript-eslint/eslint-plugin, @typescript-eslint/parser, eslint, eslint-config-prettier, eslint-plugin-prettier, prettier, typescript, tmp
[DATA] Setting resolutions: typescript@^5.0.4
[SUCCESS] Install dependencies
"
`;

exports[`Task: dependency-install > install required dependencies 1`] = `
"[STARTED] Install dependencies
[DATA] Installing devDependencies: @types/node, @typescript-eslint/eslint-plugin, @typescript-eslint/parser, eslint, eslint-config-prettier, eslint-plugin-prettier, prettier, typescript
[DATA] Setting resolutions: typescript@^5.0.4
[SUCCESS] Install dependencies
"
`;

exports[`Task: dependency-install > multiple postInstall commands from user config 1`] = `
"[STARTED] Install dependencies
[DATA] Installing devDependencies: @types/node, @typescript-eslint/eslint-plugin, @typescript-eslint/parser, eslint, eslint-config-prettier, eslint-plugin-prettier, prettier, typescript
[DATA] Run postInstall hook from config
[DATA] Setting resolutions: typescript@^5.0.4
[SUCCESS] Install dependencies
"
`;

exports[`Task: dependency-install > show failed deps 1`] = `
"We ran into an error when installing dependencies, please install the following as dependencies and try again.
  - this-is-not-existed
Error: Error: Command failed with exit code 126: volta run npm install this-is-not-existed --ignore-scripts
npm ERR! code E404
npm ERR! 404 Not Found - GET https://registry.npmjs.org/this-is-not-existed - Not found
npm ERR! 404 
npm ERR! 404  'this-is-not-existed@latest' is not in the npm registry.
npm ERR! 404 You should bug the author to publish it (or use the name yourself!)
npm ERR! 404 
npm ERR! 404 Note that you can also install from a
npm ERR! 404 tarball, folder, http url, or git url.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/malynch/.npm/_logs/2023-04-13T14_33_52_095Z-debug.log
We ran into an error when installing devDependencies, please install the following as devDependencies and try again.
  - @types/node
  - @typescript-eslint/eslint-plugin
  - @typescript-eslint/parser
  - eslint
  - eslint-config-prettier
  - eslint-plugin-prettier
  - prettier
  - typescript
  - this-would-fail
Error: Error: Command failed with exit code 126: volta run npm install @types/node @typescript-eslint/eslint-plugin @typescript-eslint/parser eslint eslint-config-prettier eslint-plugin-prettier prettier typescript this-would-fail --save-dev --ignore-scripts
npm ERR! code E404
npm ERR! 404 Not Found - GET https://registry.npmjs.org/this-would-fail - Not found
npm ERR! 404 
npm ERR! 404  'this-would-fail@latest' is not in the npm registry.
npm ERR! 404 You should bug the author to publish it (or use the name yourself!)
npm ERR! 404 
npm ERR! 404 Note that you can also install from a
npm ERR! 404 tarball, folder, http url, or git url.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/malynch/.npm/_logs/2023-04-13T14_33_54_369Z-debug.log"
`;

exports[`Task: dependency-install > single postInstall command from user config 1`] = `
"[STARTED] Install dependencies
[DATA] Installing devDependencies: @types/node, @typescript-eslint/eslint-plugin, @typescript-eslint/parser, eslint, eslint-config-prettier, eslint-plugin-prettier, prettier, typescript, fs-extra
[DATA] Run postInstall hook from config
[DATA] Setting resolutions: typescript@^5.0.4
[SUCCESS] Install dependencies
"
`;

exports[`Task: dependency-install > skip install custom deps 1`] = `
"[STARTED] Install dependencies
[SKIPPED] Install dependencies
"
`;

exports[`Task: dependency-install > skip install required dependencies 1`] = `
"[STARTED] Install dependencies
[SKIPPED] Install dependencies
"
`;
